File: Graph/C++/17086 아기 상어.cpp
#include <bits/stdc++.h>

using namespace std;

int w, h;
int board[51][51];
int visited[51][51];
int dx[8] = {-1, -1, 0, 1, 1, 1, 0, -1};
int dy[8] = {0, 1, 1, 1, 0, -1, -1, -1};
vector<pair<int, int>> sharkIndex;

int bfs(pair<int, int> index) {
    
    int safeDistance = 0;
    bool danger = false;
    queue<pair<int,int>> queue;
    queue.push(index);
    visited[index.first][index.second] = 0;
    
    while(!queue.empty()) {
        pair<int,int> v = queue.front();
        queue.pop();
        
        for(int i = 0; i < 8; i++) {
            
            int nx = v.first + dx[i];
            int ny = v.second + dy[i];
            
            if(nx < 0 || ny < 0 || nx >= h || ny >= w) continue;
            if(visited[nx][ny] > 0 || (nx == index.first && ny == index.second)) continue;
            if(board[nx][ny] == 1) {
                danger = true;
            }
            visited[nx][ny] = visited[v.first][v.second] + 1;
            safeDistance = max(safeDistance, visited[nx][ny]);
            queue.push(make_pair(nx, ny));
        }
        
        if(danger) break;
    }
    return safeDistance-1;
}

void init() {
    for(int i = 0 ; i < 51; i++) {
        fill_n(visited[i], 51, 0);
    }
}

void print() {
    
    for(int i = 0 ; i < h; i++) {
        for(int j = 0; j < w; j++) {
            cout << visited[i][j] << " ";
        }
        cout << '\n';
    }
    
    cout << '\n';
}

int main() {
    cin >> h >> w;
    for(int i = 0 ; i < h; i++) {
        for(int j = 0 ; j < w; j++) {
            cin >> board[i][j];
            if(board[i][j] == 1) sharkIndex.push_back(make_pair(i,j));
        }
    }
    
    int max_value = 0;
    
    for(int i = 0; i < sharkIndex.size(); i++) {
        max_value = max(max_value, bfs(make_pair(sharkIndex[i].first,sharkIndex[i].second)));
        print();
        init();
    }
    cout << max_value;
}

File: Python/17086 아기 상어.py
from collections import deque

dx = [-1, -1, 0, 1, 1, 1, 0, -1]
dy = [0, 1, 1, 1, 0, -1, -1, -1]
n,m = map(int, input().split())
board = []
visited = [[0 for i in range(m)] for j in range(n)]

for i in range(n) :
    row = list(map(int, input().split()))
    board.append(row)

def bfs(start) :
    if board[start[0]][start[1]] == 1 : return 0
    
    safeDistance = 0
    queue = deque()
    visited[start[0]][start[1]] = 1
    queue.append(start)
    danger = False
    while queue :
        v = queue.popleft()
        
        for i in range(8) :
            nx = dx[i] + v[0]
            ny = dy[i] + v[1]
            
            if nx not in range(n) or ny not in range(m) :
                continue
            if nx == start[0] and ny == start[1] :
                continue
            if board[nx][ny] == 1 :
                danger = True
            if visited[nx][ny] == 0 :
                visited[nx][ny] = visited[v[0]][v[1]] + 1
                safeDistance = max(safeDistance, visited[nx][ny])
                queue.append((nx, ny))
            
        if danger : break
    
    return safeDistance - 1
    
answer = 0
for i in range(n) :
    for j in range(m) :
        answer = max(answer, bfs((i,j)))
        visited = [[0 for i in range(m)] for j in range(n)]

print(answer)
